### PR TITLE
feat!: Define and implement SUM/COUNT aggregates

### PR DESCRIPTION
**Pull request by scovich**
## What changes are proposed in this pull request?

Now that we have a generalized agg evaluator in the sync engine, define and implement SUM, COUNT(x) and COUNT(*) aggs. We'll need them eventually for things like rowid base calculations and CRC computations.

### This PR affects the following public APIs

New agg types added.

## How was this change tested?

New unit tests.

### CODE REVIEW

File: kernel/src/plans/ir/nodes.rs [hunk 1]
```diff
@@ -699,22 +742,33 @@ impl Agg {
         input_schema: &StructType,
         alias: Option<String>,
     ) -> DeltaResult<StructField> {
-        // Every aggregate is nullable and preserves its value column's type and metadata.
-        let resolve = |value: &ColumnName| {
+        let resolve = |value: &ColumnName, data_type: Option<DataType>, nullable: bool| {
```
Comment (kyli87) [inline kernel/src/plans/ir/nodes.rs]: nit: maybe worth renaming `data_type` to something like `output_data_type` and then a small comment that says `None` preserves the field's data type. On first read, I thought this was specifying the input datatype that we want to preserve.

File: kernel/src/engine/sync/aggs.rs
Comment (kyli87) [inline kernel/src/engine/sync/aggs.rs:525]: Maybe I missed it, but might be worth adding a test case for sum/count over rows where all of the values are `null`?

File: kernel/proto/plan.proto [hunk 2]
```diff
@@ -92,6 +92,13 @@ message MinAgg {
 message MaxAgg {
   delta.kernel.expressions.ColumnName value = 1;
 }
+message SumAgg {
+  delta.kernel.expressions.ColumnName value = 1;
+}
+message CountAgg {
+  delta.kernel.expressions.ColumnName value = 1;
+}
+message CountStarAgg {}
```
Comment (kyli87) [inline kernel/proto/plan.proto:100]: Not opposed to having separate CountAgg and CountStar - but wondering what your main rationale for that was? Would have thought we could just have a single count and make ColumnName optional

**Review by kyli87 — APPROVED** [review]
Mostly nits, LGTM!

**Review by OussamaSaoudi — APPROVED** [review]

File: kernel/src/engine/sync/aggs.rs [hunk 3]
```diff
@@ -399,13 +473,19 @@ mod tests {
...
 
-        assert_eq!(results["a"], (Some("high"), Some("low"), Some(3), Some(1)));
-        assert_eq!(results["b"], (None, None, None, None));
+        assert_eq!(
+            results["a"],
+            (Some("high"), Some("low"), Some(3), Some(1), Some(4), 2, 2)
+        );
+        assert_eq!(results["b"], (None, None, None, None, None, 0, 1));
         Ok(())
     }
```
Comment (OussamaSaoudi) [inline kernel/src/engine/sync/aggs.rs]: Mainly I'd like tests that use assert_eq_sorted_batches for readable tests. Then execute on:
* null
* empty
* some null
* all non-null
* agg over struct (for count)

File: kernel/proto/plan.proto (hunk 2 above)
Comment (scovich) [reply kernel/proto/plan.proto:100]: We _could_ combine them but the schema derivation for `CountStar` is unique (no input columns to inherit names from) and unwrapping an option to deal with that was polluting general agg logic.

File: kernel/src/engine/sync/aggs.rs (hunk 3 above)
Comment (scovich) [reply kernel/src/engine/sync/aggs.rs]: Done!

File: kernel/src/engine/sync/aggs.rs
Comment (scovich) [reply kernel/src/engine/sync/aggs.rs:525]: Done!

File: kernel/src/plans/ir/nodes.rs (hunk 1 above)
Comment (scovich) [reply kernel/src/plans/ir/nodes.rs]: Done
